// @superfaceai/one-sdk/dist/core/profile-provider/bound-profile-provider.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BoundProfileProvider = void 0;
var lib_1 = require("../../lib");
var events_1 = require("../events");
var interpreter_1 = require("../interpreter");
var parameters_1 = require("./parameters");
var security_1 = require("./security");
var DEBUG_NAMESPACE_SENSITIVE = 'bound-profile-provider:sensitive';
var BoundProfileProvider = /** @class */ (function () {
    function BoundProfileProvider(profileAst, mapAst, provider, config, configuration, crypto, fetchInstance, logger, events) {
        this.profileAst = profileAst;
        this.mapAst = mapAst;
        this.provider = provider;
        this.config = config;
        this.configuration = configuration;
        this.crypto = crypto;
        this.fetchInstance = fetchInstance;
        this.logger = logger;
        this.profileValidator = new interpreter_1.ProfileParameterValidator(this.profileAst, this.logger);
        this.fetchInstance.metadata = {
            profile: lib_1.profileAstId(profileAst),
            provider: provider.name,
        };
        this.fetchInstance.events = events;
        this.logSensitive = logger === null || logger === void 0 ? void 0 : logger.log(DEBUG_NAMESPACE_SENSITIVE);
    }
    /**
     * Performs the usecase while validating input and output against the profile definition.
     *
     * Note that the `TInput` and `TResult` types cannot be checked for compatibility with the profile definition, so the caller
     * is responsible for ensuring that the cast is safe.
     */
    BoundProfileProvider.prototype.perform = function (usecase, input, parameters, securityValues) {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var composedInput, inputValidation, security, interpreter, result, resultValidation;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        this.fetchInstance.metadata = {
                            profile: lib_1.profileAstId(this.profileAst),
                            usecase: usecase,
                            provider: this.provider.name,
                        };
                        composedInput = this.composeInput(usecase, input);
                        inputValidation = this.profileValidator.validate(composedInput, 'input', usecase);
                        if (inputValidation.isErr()) {
                            return [2 /*return*/, lib_1.err(inputValidation.error)];
                        }
                        lib_1.forceCast(composedInput);
                        security = securityValues
                            ? security_1.resolveSecurityConfiguration((_a = this.provider.securitySchemes) !== null && _a !== void 0 ? _a : [], securityValues, this.provider.name)
                            : this.configuration.security;
                        interpreter = new interpreter_1.MapInterpreter({
                            input: composedInput,
                            usecase: usecase,
                            services: this.configuration.services,
                            security: security,
                            parameters: parameters_1.resolveIntegrationParameters(this.provider, this.mergeParameters(parameters, this.configuration.parameters)),
                        }, {
                            config: this.config,
                            fetchInstance: this.fetchInstance,
                            externalHandler: new events_1.MapInterpreterEventAdapter(this.fetchInstance.metadata, this.fetchInstance.events),
                            logger: this.logger,
                            crypto: this.crypto,
                        });
                        return [4 /*yield*/, interpreter.perform(this.mapAst)];
                    case 1:
                        result = _b.sent();
                        if (result.isErr()) {
                            return [2 /*return*/, lib_1.err(result.error)];
                        }
                        resultValidation = this.profileValidator.validate(result.value, 'result', usecase);
                        if (resultValidation.isErr()) {
                            return [2 /*return*/, lib_1.err(resultValidation.error)];
                        }
                        lib_1.forceCast(result.value);
                        return [2 /*return*/, lib_1.ok(result.value)];
                }
            });
        });
    };
    BoundProfileProvider.prototype.composeInput = function (usecase, input) {
        var _a, _b, _c;
        var composed = input;
        var defaultInput = (_b = (_a = this.configuration.profileProviderSettings) === null || _a === void 0 ? void 0 : _a.defaults[usecase]) === null || _b === void 0 ? void 0 : _b.input;
        if (defaultInput !== undefined) {
            composed = lib_1.mergeVariables(lib_1.castToNonPrimitive(defaultInput), input !== null && input !== void 0 ? input : {});
            (_c = this.logSensitive) === null || _c === void 0 ? void 0 : _c.call(this, 'Composed input with defaults: %O', composed);
        }
        return composed;
    };
    BoundProfileProvider.prototype.mergeParameters = function (parameters, providerParameters) {
        if (parameters === undefined) {
            return providerParameters;
        }
        if (providerParameters === undefined) {
            return parameters;
        }
        return __assign(__assign({}, providerParameters), parameters);
    };
    return BoundProfileProvider;
}());
exports.BoundProfileProvider = BoundProfileProvider;
