// @superfaceai/ast/dist/interfaces/ast/document.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isUnknownFile = exports.isMapASTFile = exports.isMapFile = exports.isProfileASTFile = exports.isProfileFile = exports.inferDocumentType = exports.DocumentType = exports.EXTENSIONS = void 0;
exports.EXTENSIONS = {
    profile: {
        source: '.supr',
        build: '.supr.ast.json',
    },
    map: {
        source: '.suma',
        build: '.suma.ast.json',
    },
};
var DocumentType;
(function (DocumentType) {
    DocumentType["UNKNOWN"] = "unknown";
    DocumentType["MAP"] = "map";
    DocumentType["PROFILE"] = "profile";
    DocumentType["MAP_AST"] = "map.ast";
    DocumentType["PROFILE_AST"] = "profile.ast";
})(DocumentType = exports.DocumentType || (exports.DocumentType = {}));
/**
 * Detects whether the file on path is Superface Map or Superface Profile based on the extension.
 */
function inferDocumentType(path) {
    var normalizedPath = path.toLowerCase().trim();
    if (normalizedPath.endsWith(exports.EXTENSIONS.map.source)) {
        return "map" /* MAP */;
    }
    if (normalizedPath.endsWith(exports.EXTENSIONS.profile.source)) {
        return "profile" /* PROFILE */;
    }
    if (normalizedPath.endsWith(exports.EXTENSIONS.map.build)) {
        return "map.ast" /* MAP_AST */;
    }
    if (normalizedPath.endsWith(exports.EXTENSIONS.profile.build)) {
        return "profile.ast" /* PROFILE_AST */;
    }
    return "unknown" /* UNKNOWN */;
}
exports.inferDocumentType = inferDocumentType;
/**
 * Checks whether the file is a Profile source file
 **/
function isProfileFile(file) {
    return inferDocumentType(file) === "profile" /* PROFILE */;
}
exports.isProfileFile = isProfileFile;
/**
 * Checks whether the file is a compiled Profile file
 **/
function isProfileASTFile(file) {
    return inferDocumentType(file) === "profile.ast" /* PROFILE_AST */;
}
exports.isProfileASTFile = isProfileASTFile;
/**
 * Checks whether the file is a Map source file
 **/
function isMapFile(file) {
    return inferDocumentType(file) === "map" /* MAP */;
}
exports.isMapFile = isMapFile;
/**
 * Checks whether the file is a compiled Map file
 **/
function isMapASTFile(file) {
    return inferDocumentType(file) === "map.ast" /* MAP_AST */;
}
exports.isMapASTFile = isMapASTFile;
/**
 * Checks whether the file is neither map nor profile file
 **/
function isUnknownFile(file) {
    return inferDocumentType(file) === "unknown" /* UNKNOWN */;
}
exports.isUnknownFile = isUnknownFile;
