// @superfaceai/ast/dist/interfaces/providerjson/providerjson.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpScheme = exports.ApiKeyPlacement = exports.SecurityType = exports.PROVIDER_NAME_REGEX_SOURCE = exports.PROVIDER_NAME_REGEX = void 0;
exports.PROVIDER_NAME_REGEX = new RegExp('^[a-z][_\\-a-z]*$');
exports.PROVIDER_NAME_REGEX_SOURCE = exports.PROVIDER_NAME_REGEX.source;
/**
 * Type of security value.
 */
var SecurityType;
(function (SecurityType) {
    SecurityType["APIKEY"] = "apiKey";
    SecurityType["HTTP"] = "http";
})(SecurityType = exports.SecurityType || (exports.SecurityType = {}));
/**
 * The placement of the API key.
 * @$id ApiKeyPlacement
 **/
var ApiKeyPlacement;
(function (ApiKeyPlacement) {
    ApiKeyPlacement["HEADER"] = "header";
    ApiKeyPlacement["BODY"] = "body";
    ApiKeyPlacement["PATH"] = "path";
    ApiKeyPlacement["QUERY"] = "query";
})(ApiKeyPlacement = exports.ApiKeyPlacement || (exports.ApiKeyPlacement = {}));
var HttpScheme;
(function (HttpScheme) {
    HttpScheme["BASIC"] = "basic";
    HttpScheme["BEARER"] = "bearer";
    HttpScheme["DIGEST"] = "digest";
})(HttpScheme = exports.HttpScheme || (exports.HttpScheme = {}));
