// @superfaceai/one-sdk/dist/schema-tools/superjson/normalize.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizeSuperJsonDocument = exports.normalizeProviderSettings = exports.normalizeProfileSettings = exports.normalizeProfileProviderDefaults = exports.normalizeUsecaseDefaults = exports.normalizeRetryPolicy = exports.normalizeBackoff = exports.normalizeProfileProviderSettings = void 0;
var ast_1 = require("@superfaceai/ast");
var lib_1 = require("../../lib");
var errors_helpers_1 = require("./errors.helpers");
function normalizeProfileProviderSettings(profileProviderSettings, baseDefaults, environment, logger) {
    if (profileProviderSettings === undefined) {
        return {
            defaults: {},
        };
    }
    if (typeof profileProviderSettings === 'string') {
        if (ast_1.isFileURIString(profileProviderSettings)) {
            return {
                file: profileProviderSettings.slice(ast_1.FILE_URI_PROTOCOL.length),
                defaults: {},
            };
        }
        throw errors_helpers_1.invalidProfileProviderError(profileProviderSettings);
    }
    var normalizedSettings;
    if ('file' in profileProviderSettings) {
        normalizedSettings = {
            file: profileProviderSettings.file,
            defaults: {},
        };
    }
    else {
        normalizedSettings = {
            mapVariant: profileProviderSettings.mapVariant,
            mapRevision: profileProviderSettings.mapRevision,
            defaults: {},
        };
    }
    normalizedSettings.defaults = normalizeProfileProviderDefaults(environment, logger, profileProviderSettings.defaults, baseDefaults);
    return normalizedSettings;
}
exports.normalizeProfileProviderSettings = normalizeProfileProviderSettings;
function normalizeBackoff(backoff, base) {
    var _a, _b;
    if (backoff === undefined) {
        if (base === undefined) {
            return { kind: ast_1.BackoffKind.EXPONENTIAL };
        }
        else {
            return normalizeBackoff(base);
        }
    }
    // string-to-object expansion
    var objectBackoff;
    if (backoff === ast_1.BackoffKind.EXPONENTIAL) {
        objectBackoff = { kind: ast_1.BackoffKind.EXPONENTIAL };
    }
    else {
        objectBackoff = backoff;
    }
    return {
        kind: ast_1.BackoffKind.EXPONENTIAL,
        start: (_a = objectBackoff.start) !== null && _a !== void 0 ? _a : base === null || base === void 0 ? void 0 : base.start,
        factor: (_b = objectBackoff.factor) !== null && _b !== void 0 ? _b : base === null || base === void 0 ? void 0 : base.factor,
    };
}
exports.normalizeBackoff = normalizeBackoff;
function normalizeRetryPolicy(retryPolicy, base) {
    var _a, _b, _c, _d, _e;
    if (retryPolicy === undefined) {
        if (base === undefined) {
            return { kind: ast_1.OnFail.NONE };
        }
        else {
            return normalizeRetryPolicy(base);
        }
    }
    // string-to-object expansion - always fully override base
    var objectRetryPolicy;
    if (retryPolicy === ast_1.OnFail.CIRCUIT_BREAKER) {
        objectRetryPolicy = { kind: ast_1.OnFail.CIRCUIT_BREAKER };
    }
    else if (retryPolicy === ast_1.OnFail.SIMPLE) {
        objectRetryPolicy = { kind: ast_1.OnFail.SIMPLE };
    }
    else if (retryPolicy === ast_1.OnFail.NONE) {
        objectRetryPolicy = { kind: ast_1.OnFail.NONE };
    }
    else {
        objectRetryPolicy = retryPolicy;
    }
    if (objectRetryPolicy.kind === ast_1.OnFail.SIMPLE) {
        if ((base === null || base === void 0 ? void 0 : base.kind) === ast_1.OnFail.SIMPLE) {
            (_a = objectRetryPolicy.maxContiguousRetries) !== null && _a !== void 0 ? _a : (objectRetryPolicy.maxContiguousRetries = base.maxContiguousRetries);
            (_b = objectRetryPolicy.requestTimeout) !== null && _b !== void 0 ? _b : (objectRetryPolicy.requestTimeout = base.requestTimeout);
        }
        return objectRetryPolicy;
    }
    if (objectRetryPolicy.kind === ast_1.OnFail.CIRCUIT_BREAKER) {
        var normalizedBackoff = void 0;
        if ((base === null || base === void 0 ? void 0 : base.kind) === ast_1.OnFail.CIRCUIT_BREAKER) {
            (_c = objectRetryPolicy.maxContiguousRetries) !== null && _c !== void 0 ? _c : (objectRetryPolicy.maxContiguousRetries = base.maxContiguousRetries);
            (_d = objectRetryPolicy.requestTimeout) !== null && _d !== void 0 ? _d : (objectRetryPolicy.requestTimeout = base.requestTimeout);
            (_e = objectRetryPolicy.openTime) !== null && _e !== void 0 ? _e : (objectRetryPolicy.openTime = base.openTime);
            normalizedBackoff = normalizeBackoff(objectRetryPolicy.backoff, base.backoff);
        }
        else {
            normalizedBackoff = normalizeBackoff(objectRetryPolicy.backoff);
        }
        return __assign(__assign({}, objectRetryPolicy), { backoff: normalizedBackoff });
    }
    return objectRetryPolicy;
}
exports.normalizeRetryPolicy = normalizeRetryPolicy;
function normalizeUsecaseDefaults(environment, logger, defaults, base) {
    var e_1, _a;
    var _b, _c, _d, _e;
    if (defaults === undefined) {
        if (base == undefined) {
            return {};
        }
        else {
            return normalizeUsecaseDefaults(environment, logger, base);
        }
    }
    var normalized = base !== undefined ? lib_1.clone(base) : {};
    try {
        for (var _f = __values(Object.entries(defaults)), _g = _f.next(); !_g.done; _g = _f.next()) {
            var _h = __read(_g.value, 2), usecase = _h[0], defs = _h[1];
            var previousInput = lib_1.castToNonPrimitive((_c = (_b = normalized[usecase]) === null || _b === void 0 ? void 0 : _b.input) !== null && _c !== void 0 ? _c : {});
            var providerFailover = defs.providerFailover;
            if (providerFailover === undefined) {
                providerFailover = (_d = normalized[usecase]) === null || _d === void 0 ? void 0 : _d.providerFailover;
            }
            normalized[usecase] = {
                input: lib_1.mergeVariables(previousInput, lib_1.castToNonPrimitive((_e = defs.input) !== null && _e !== void 0 ? _e : {})),
                providerFailover: providerFailover !== null && providerFailover !== void 0 ? providerFailover : false,
            };
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (_g && !_g.done && (_a = _f.return)) _a.call(_f);
        }
        finally { if (e_1) throw e_1.error; }
    }
    if (environment !== undefined) {
        return lib_1.resolveEnvRecord(normalized, environment, logger);
    }
    return normalized;
}
exports.normalizeUsecaseDefaults = normalizeUsecaseDefaults;
function normalizeProfileProviderDefaults(environment, logger, defaults, base) {
    var e_2, _a;
    var _b, _c, _d;
    if (defaults === undefined) {
        if (base == undefined) {
            return {};
        }
        else {
            return normalizeProfileProviderDefaults(environment, logger, base);
        }
    }
    var normalized = {};
    try {
        for (var _e = __values(Object.entries(defaults)), _f = _e.next(); !_f.done; _f = _e.next()) {
            var _g = __read(_f.value, 2), usecase = _g[0], defs = _g[1];
            var previousInput = lib_1.castToNonPrimitive((_c = (_b = base === null || base === void 0 ? void 0 : base[usecase]) === null || _b === void 0 ? void 0 : _b.input) !== null && _c !== void 0 ? _c : {});
            normalized[usecase] = {
                input: lib_1.mergeVariables(previousInput, lib_1.castToNonPrimitive((_d = defs.input) !== null && _d !== void 0 ? _d : {})),
                retryPolicy: normalizeRetryPolicy(defs.retryPolicy),
            };
        }
    }
    catch (e_2_1) { e_2 = { error: e_2_1 }; }
    finally {
        try {
            if (_f && !_f.done && (_a = _e.return)) _a.call(_e);
        }
        finally { if (e_2) throw e_2.error; }
    }
    if (environment !== undefined) {
        return lib_1.resolveEnvRecord(normalized, environment, logger);
    }
    return normalized;
}
exports.normalizeProfileProviderDefaults = normalizeProfileProviderDefaults;
function normalizeProfileSettings(profileEntry, topProviderOrder, environment, logger) {
    var e_3, _a;
    var _b;
    if (typeof profileEntry === 'string') {
        if (ast_1.isVersionString(profileEntry)) {
            return {
                version: profileEntry,
                priority: topProviderOrder,
                defaults: {},
                providers: {},
            };
        }
        if (ast_1.isFileURIString(profileEntry)) {
            return {
                file: profileEntry.slice(ast_1.FILE_URI_PROTOCOL.length),
                priority: topProviderOrder,
                defaults: {},
                providers: {},
            };
        }
        throw new lib_1.UnexpectedError('Invalid profile entry format: ' + profileEntry);
    }
    var normalizedSettings;
    if ('file' in profileEntry) {
        normalizedSettings = {
            file: profileEntry.file,
            priority: profileEntry.priority || [],
            defaults: {},
            providers: {},
        };
    }
    else {
        normalizedSettings = {
            version: profileEntry.version,
            priority: profileEntry.priority || [],
            defaults: {},
            providers: {},
        };
    }
    normalizedSettings.defaults = normalizeUsecaseDefaults(environment, logger, profileEntry.defaults);
    try {
        for (var _c = __values(Object.entries((_b = profileEntry.providers) !== null && _b !== void 0 ? _b : {})), _d = _c.next(); !_d.done; _d = _c.next()) {
            var _e = __read(_d.value, 2), providerName = _e[0], profileProviderSettings = _e[1];
            normalizedSettings.providers[providerName] =
                normalizeProfileProviderSettings(profileProviderSettings, normalizedSettings.defaults, environment, logger);
        }
    }
    catch (e_3_1) { e_3 = { error: e_3_1 }; }
    finally {
        try {
            if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
        }
        finally { if (e_3) throw e_3.error; }
    }
    if (normalizedSettings.priority.length === 0) {
        var providerOrder = Object.keys(profileEntry.providers || {});
        normalizedSettings.priority =
            providerOrder.length > 0 ? providerOrder : topProviderOrder;
    }
    if (normalizedSettings.priority.length === 0) {
        var providerOrder = Object.keys(profileEntry.providers || {});
        normalizedSettings.priority =
            providerOrder.length > 0 ? providerOrder : topProviderOrder;
    }
    return normalizedSettings;
}
exports.normalizeProfileSettings = normalizeProfileSettings;
function normalizeProviderSettings(providerEntry, environment, logger) {
    var _a, _b, _c, _d;
    if (typeof providerEntry === 'string') {
        if (ast_1.isFileURIString(providerEntry)) {
            return {
                file: providerEntry.slice(ast_1.FILE_URI_PROTOCOL.length),
                security: [],
                parameters: {},
            };
        }
        throw new lib_1.UnexpectedError('Invalid provider entry format: ' + providerEntry);
    }
    if (environment !== undefined) {
        return {
            file: providerEntry.file,
            security: (_b = (_a = providerEntry.security) === null || _a === void 0 ? void 0 : _a.map(function (entry) {
                return lib_1.resolveEnvRecord(entry, environment, logger);
            })) !== null && _b !== void 0 ? _b : [],
            parameters: providerEntry.parameters
                ? lib_1.resolveEnvRecord(providerEntry.parameters, environment, logger)
                : {},
        };
    }
    return {
        file: providerEntry.file,
        security: (_c = providerEntry.security) !== null && _c !== void 0 ? _c : [],
        parameters: (_d = providerEntry.parameters) !== null && _d !== void 0 ? _d : {},
    };
}
exports.normalizeProviderSettings = normalizeProviderSettings;
/** Returns a cached normalized clone of the document. */
function normalizeSuperJsonDocument(originalDocument, environment, logger) {
    var e_4, _a, e_5, _b;
    var _c, _d, _e;
    var document = lib_1.clone(originalDocument);
    var profiles = (_c = document.profiles) !== null && _c !== void 0 ? _c : {};
    var normalizedProfiles = {};
    var topProviderOrder = Object.keys((_d = originalDocument.providers) !== null && _d !== void 0 ? _d : {});
    try {
        for (var _f = __values(Object.entries(profiles)), _g = _f.next(); !_g.done; _g = _f.next()) {
            var _h = __read(_g.value, 2), profileId = _h[0], profileEntry = _h[1];
            normalizedProfiles[profileId] = normalizeProfileSettings(profileEntry, topProviderOrder, environment, logger);
        }
    }
    catch (e_4_1) { e_4 = { error: e_4_1 }; }
    finally {
        try {
            if (_g && !_g.done && (_a = _f.return)) _a.call(_f);
        }
        finally { if (e_4) throw e_4.error; }
    }
    var providers = (_e = document.providers) !== null && _e !== void 0 ? _e : {};
    var normalizedProviders = {};
    try {
        for (var _j = __values(Object.entries(providers)), _k = _j.next(); !_k.done; _k = _j.next()) {
            var _l = __read(_k.value, 2), providerName = _l[0], providerEntry = _l[1];
            normalizedProviders[providerName] = normalizeProviderSettings(providerEntry, environment, logger);
        }
    }
    catch (e_5_1) { e_5 = { error: e_5_1 }; }
    finally {
        try {
            if (_k && !_k.done && (_b = _j.return)) _b.call(_j);
        }
        finally { if (e_5) throw e_5.error; }
    }
    return {
        profiles: normalizedProfiles,
        providers: normalizedProviders,
    };
}
exports.normalizeSuperJsonDocument = normalizeSuperJsonDocument;
