// @superfaceai/one-sdk/dist/core/interpreter/http/interfaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BINARY_CONTENT_REGEXP = exports.BINARY_CONTENT_TYPES = exports.FORMDATA_CONTENT = exports.URLENCODED_CONTENT = exports.JSON_PROBLEM_CONTENT = exports.JSON_CONTENT = exports.isBinaryBody = exports.isUrlSearchParamsBody = exports.isFormDataBody = exports.isStringBody = exports.binaryBody = exports.urlSearchParamsBody = exports.formDataBody = exports.stringBody = void 0;
var stringBody = function (data) { return ({
    _type: 'string',
    data: data,
}); };
exports.stringBody = stringBody;
var formDataBody = function (data) { return ({
    _type: 'formdata',
    data: data,
}); };
exports.formDataBody = formDataBody;
var urlSearchParamsBody = function (data) { return ({ _type: 'urlsearchparams', data: data }); };
exports.urlSearchParamsBody = urlSearchParamsBody;
var binaryBody = function (data) { return ({
    _type: 'binary',
    data: data,
}); };
exports.binaryBody = binaryBody;
function isStringBody(data) {
    return data._type === 'string';
}
exports.isStringBody = isStringBody;
function isFormDataBody(data) {
    return data._type === 'formdata';
}
exports.isFormDataBody = isFormDataBody;
function isUrlSearchParamsBody(data) {
    return data._type === 'urlsearchparams';
}
exports.isUrlSearchParamsBody = isUrlSearchParamsBody;
function isBinaryBody(data) {
    return data._type === 'binary';
}
exports.isBinaryBody = isBinaryBody;
exports.JSON_CONTENT = 'application/json';
exports.JSON_PROBLEM_CONTENT = 'application/problem+json';
exports.URLENCODED_CONTENT = 'application/x-www-form-urlencoded';
exports.FORMDATA_CONTENT = 'multipart/form-data';
exports.BINARY_CONTENT_TYPES = [
    'application/octet-stream',
    'video/*',
    'audio/*',
    'image/*',
];
exports.BINARY_CONTENT_REGEXP = /application\/octet-stream|video\/.*|audio\/.*|image\/.*/;
