// @superfaceai/ast/dist/interfaces/superjson/superjson.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BackoffKind = exports.OnFail = exports.FILE_URI_REGEX_SOURCE = exports.FILE_URI_REGEX = exports.FILE_URI_PROTOCOL = exports.SEMVER_REGEX_SOURCE = exports.SEMVER_REGEX = void 0;
// 'Official' regex https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
exports.SEMVER_REGEX = /^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$/;
exports.SEMVER_REGEX_SOURCE = exports.SEMVER_REGEX.source;
// NOT comprehensive at all
exports.FILE_URI_PROTOCOL = 'file://';
exports.FILE_URI_REGEX = /^file:\/\//;
exports.FILE_URI_REGEX_SOURCE = exports.FILE_URI_REGEX.source;
// Retry policy
var OnFail;
(function (OnFail) {
    OnFail["NONE"] = "none";
    OnFail["SIMPLE"] = "simple";
    OnFail["CIRCUIT_BREAKER"] = "circuit-breaker";
})(OnFail = exports.OnFail || (exports.OnFail = {}));
var BackoffKind;
(function (BackoffKind) {
    BackoffKind["EXPONENTIAL"] = "exponential";
})(BackoffKind = exports.BackoffKind || (exports.BackoffKind = {}));
