// @superfaceai/one-sdk/dist/core/events/failure/backoff.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExponentialBackoff = exports.LinearBackoff = exports.ConstantBackoff = exports.Backoff = void 0;
var Backoff = /** @class */ (function () {
    function Backoff(successor, inverseSuccessor, initial) {
        this.successor = successor;
        this.inverseSuccessor = inverseSuccessor;
        this._current = initial;
    }
    Object.defineProperty(Backoff.prototype, "current", {
        get: function () {
            return this._current;
        },
        enumerable: false,
        configurable: true
    });
    Backoff.prototype.up = function () {
        this._current = this.successor(this._current);
        return this._current;
    };
    Backoff.prototype.down = function () {
        this._current = this.inverseSuccessor(this._current);
        return this._current;
    };
    Backoff.clampValue = function (value, minimum, maximum) {
        if (minimum !== undefined) {
            value = Math.max(minimum, value);
        }
        if (maximum !== undefined) {
            value = Math.min(maximum, value);
        }
        return value;
    };
    Backoff.DEFAULT_INITIAL = 500;
    return Backoff;
}());
exports.Backoff = Backoff;
var ConstantBackoff = /** @class */ (function (_super) {
    __extends(ConstantBackoff, _super);
    function ConstantBackoff(initial) {
        return _super.call(this, function (x) { return x; }, function (x) { return x; }, initial) || this;
    }
    return ConstantBackoff;
}(Backoff));
exports.ConstantBackoff = ConstantBackoff;
var LinearBackoff = /** @class */ (function (_super) {
    __extends(LinearBackoff, _super);
    function LinearBackoff(initial, step, minimum, maximum) {
        return _super.call(this, function (x) { return Backoff.clampValue(x + step, undefined, maximum); }, function (x) { return Backoff.clampValue(x - step, minimum, undefined); }, initial) || this;
    }
    return LinearBackoff;
}(Backoff));
exports.LinearBackoff = LinearBackoff;
var ExponentialBackoff = /** @class */ (function (_super) {
    __extends(ExponentialBackoff, _super);
    function ExponentialBackoff(initial, base, minimum, maximum) {
        return _super.call(this, function (x) { return Backoff.clampValue(x * base, undefined, maximum); }, function (x) { return Backoff.clampValue(x / base, minimum, undefined); }, initial) || this;
    }
    ExponentialBackoff.DEFAULT_BASE = 2;
    return ExponentialBackoff;
}(Backoff));
exports.ExponentialBackoff = ExponentialBackoff;
